[FLINK-26214] Publish Docker images to GitHub Registry

mbalassi opened the pull request:
I managed to make this work with simply using `flink-operator` for the docker image name, after reconsideration I prefer this for brevity.
We only publish images for the `main` branch after this PR, I will proceed with working on the release branch/tag setup.
We should improve this ideally where we only tag the image with the `latest` tag if the CI finishes successfully.

mbalassi commented:
I tested via [mbalassi/packages](https://github.com/mbalassi/flink-kubernetes-operator/pkgs/container/flink-operator), and could already manually push to [apache/packages](https://github.com/orgs/apache/packages/container/package/flink-operator).

tisonkun commented on .github/workflows/docker_push.yml:
```diff
@@ -0,0 +1,59 @@
+        uses: actions/checkout@v2
+
+      - name: Log in to the Container registry
+        uses: docker/login-action@f054a8b539a109f9f41c372932f1ae047eff08c9
```
```suggestion
        uses: docker/login-action@v1
```

mbalassi replied:
Good call, I cleaned up references like this

tisonkun commented on .github/workflows/docker_push.yml:
```diff
@@ -0,0 +1,59 @@
+          context: .
+          push: true
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
```
```suggestion
          labels: ${{ steps.meta.outputs.labels }}

```

tisonkun commented on .github/workflows/docker_push.yml:
```diff
@@ -0,0 +1,59 @@
+        uses: docker/build-push-action@ad44023a93711e3deb337508980b4b5e9bcdc5dc
+        with:
+          context: .
+          push: true
```
You may trigger the docker packaging logic on `pull_request` also while skip the push phase so that we can check before merging a pull request. You can refer to [this workflow](https://github.com/engula/engula/blob/93cd8b3be0cd8ffc6a3a7f331e6bde2ca9d15284/.github/workflows/docker.yml) as an example.

mbalassi replied:
Great!

tisonkun reviewed (commented):
Generally looks good! Comments inline.

mbalassi commented:
Thanks, @tisonkun I addressed your comments. I went with an approach based on the [Semantic versioning](https://github.com/docker/metadata-action#semver) example of the docker metadata action.

gyfora commented on .github/workflows/docker_push.yml:
```diff
@@ -0,0 +1,67 @@
+      - main
+      - 'release-*'
+    tags:
+      - 'v*'
```
this should be `release-*

mbalassi replied:
Good catch

gyfora commented on .github/workflows/docker_push.yml:
```diff
@@ -0,0 +1,67 @@
+          images: |
+            ghcr.io/${{ github.repository_owner }}/flink-operator
+          tags: |
+            type=raw,value=main,enable=${{ github.ref == 'refs/heads/main') }}
```
remove extra `)` please
